File: Hivemind/Hivemind/app/services/providers/error_classifier.rb
# frozen_string_literal: true

module Providers
  # The one place a provider failure becomes a typed verdict.
  #
  # Mirrors sdk-proxy/error-classifier.js so both sides of the proxy agree on
  # what is permanent. When the proxy sends a structured body we trust its
  # verdict verbatim; otherwise (direct API path, Faraday, OpenAI, Ollama) we
  # derive the same answer from status and message here.
  #
  # Deliberate default: anything unclassifiable is PERMANENT. Treating the
  # unknown as retryable is precisely what turned one out-of-credit account
  # into a host-wide ephemeral-port exhaustion on 2026-08-24.
  class ErrorClassifier
    QUOTA_PATTERNS = /
      out\ of\ extra\ usage |
      credit\ balance\ is\ too\ low |
      insufficient[_\s-]?quota |
      billing[_\s-]?(?:hard[_\s-]?)?limit |
      purchase\ (?:more\ )?credits |
      claude\.ai\/settings\/usage
    /xi

    AUTH_PATTERNS = /
      invalid[_\s-]?api[_\s-]?key | authentication[_\s-]?error |
      oauth\ token\ (?:has\ )?expired | invalid\ bearer\ token | unauthorized
    /xi

    MODEL_NOT_FOUND_PATTERNS = /not_found_error | unknown\ model | model[:\s].*not[_\s-]?found/xi

    # Local resource exhaustion. NOT retryable: EADDRNOTAVAIL means the kernel
    # has no ephemeral port left, and every retry makes that worse.
    LOCAL_EXHAUSTION_PATTERNS = /
      EADDRNOTAVAIL | can't\ assign\ requested\ address |
      EMFILE | ENFILE | too\ many\ open\ files
    /xi

    NETWORK_PATTERNS = /
      ECONNRESET | ECONNREFUSED | EPIPE | ENOTFOUND | EAI_AGAIN |
      ETIMEDOUT | ESOCKETTIMEDOUT | socket\ hang\ up | execution\ expired |
      network[_\s-]?error | connection\ (?:error|refused|reset|failed) |
      timed[_\s-]?out | timeout | failed\ to\ open\ tcp
    /xi

    OVERLOAD_PATTERNS = /overloaded | rate[_\s-]?limit | too\ many\ requests/xi

    STATUS_IN_TEXT = /
      \(\s*(\d{3})\s*\) |
      \bAPI\ Error:?\s*(\d{3})\b |
      \bstatus(?:\s*code)?[:\s]+(\d{3})\b |
      \bHTTP\ (\d{3})\b
    /xi

    # reason => [permanent?, default status]
    REASONS = {
      "quota_exhausted"       => [ true,  402 ],
      "auth_invalid"          => [ true,  401 ],
      "forbidden"             => [ true,  403 ],
      "model_not_found"       => [ true,  404 ],
      "invalid_request"       => [ true,  400 ],
      "request_too_large"     => [ true,  413 ],
      "local_port_exhaustion" => [ true,  503 ],
      "unknown"               => [ true,  500 ],
      "rate_limited"          => [ false, 429 ],
      "server_error"          => [ false, 502 ],
      "network_error"         => [ false, 502 ],
      "timeout"               => [ false, 504 ],
      "conflict"              => [ false, 409 ]
    }.freeze

    # Reasons that count toward opening the credential's circuit. A malformed
    # request or a missing model is the caller's bug, not the credential's —
    # those must not silence a working account.
    CIRCUIT_REASONS = %w[quota_exhausted auth_invalid forbidden local_port_exhaustion].freeze

    class << self
      # Build a typed error from anything a provider path can produce.
      #
      # @param message  [String]      the failure text
      # @param status   [Integer,nil] HTTP status if known
      # @param body     [String,Hash,nil] response body; a structured proxy
      #                 body ({"reason":..., "retryable":...}) wins outright
      # @param provider [String,nil]
      # @return [ProviderError]
      def call(message:, status: nil, body: nil, provider: nil)
        verdict = from_structured_body(body) || derive(message.to_s, status, body)

        klass = verdict[:retryable] ? TransientProviderError : PermanentProviderError
        klass.new(
          message.to_s.presence || "Provider call failed (#{verdict[:reason]})",
          status: verdict[:status],
          reason: verdict[:reason],
          provider: provider,
          retry_after: verdict[:retry_after]
        )
      end

      # Classify a ServiceResponse failure string produced by an adapter that
      # has already flattened its error. Used by FailoverAdapter and callers
      # holding only a ServiceResponse.
      def from_error_string(error, provider: nil)
        call(message: error.to_s, status: nil, provider: provider)
      end

      def opens_circuit?(error)
        error.is_a?(ProviderError) && CIRCUIT_REASONS.include?(error.reason)
      end

      def permanent?(reason) = REASONS.fetch(reason.to_s, [ true ]).first

      private

      # The sdk-proxy already classified this; re-deriving from status would
      # be strictly worse (a 503 from an open circuit is not a server error).
      def from_structured_body(body)
        parsed = parse_body(body)
        return nil unless parsed.is_a?(Hash)

        reason = parsed["reason"].presence
        retryable = parsed["retryable"]
        return nil if reason.nil? || ![ true, false ].include?(retryable)

        {
          reason: reason,
          retryable: retryable,
          status: REASONS.dig(reason, 1) || 500,
          retry_after: retry_after_from(parsed)
        }
      end

      def parse_body(body)
        case body
        when Hash then body.stringify_keys
        when String then (JSON.parse(body) if body.strip.start_with?("{"))
        end
      rescue JSON::ParserError
        nil
      end

      def retry_after_from(parsed)
        ms = parsed["retry_after_ms"]
        ms.is_a?(Numeric) ? (ms / 1000.0).ceil : nil
      end

      def derive(message, status, body)
        haystack = [ message, body_text(body) ].compact.join("\n")
        effective_status = status || status_from_text(haystack)
        reason = reason_for_status(effective_status, haystack) || reason_from_text(haystack) || "unknown"

        {
          reason: reason,
          retryable: !permanent?(reason),
          status: effective_status || REASONS.dig(reason, 1) || 500,
          retry_after: nil
        }
      end

      def body_text(body)
        body.is_a?(String) ? body : (body&.to_json if body.respond_to?(:to_json))
      end

      def status_from_text(text)
        match = text.match(STATUS_IN_TEXT)
        return nil unless match

        code = match.captures.compact.first.to_i
        code.between?(400, 599) ? code : nil
      end

      def reason_for_status(status, text)
        case status
        when nil then nil
        when 400 then text.match?(QUOTA_PATTERNS) ? "quota_exhausted" : "invalid_request"
        when 401 then "auth_invalid"
        when 402 then "quota_exhausted"
        when 403 then "forbidden"
        when 404 then "model_not_found"
        when 408 then "timeout"
        when 409 then "conflict"
        when 413 then "request_too_large"
        when 422 then "invalid_request"
        when 429 then "rate_limited"
        when 500..599 then "server_error"
        end
      end

      def reason_from_text(text)
        # Order matters: local exhaustion and quota are checked before the
        # generic network patterns, whose messages often overlap.
        return "local_port_exhaustion" if text.match?(LOCAL_EXHAUSTION_PATTERNS)
        return "quota_exhausted"       if text.match?(QUOTA_PATTERNS)
        return "auth_invalid"          if text.match?(AUTH_PATTERNS)
        return "model_not_found"       if text.match?(MODEL_NOT_FOUND_PATTERNS)
        return "rate_limited"          if text.match?(OVERLOAD_PATTERNS)
        return "network_error"         if text.match?(NETWORK_PATTERNS)

        nil
      end
    end
  end
end

File: ➡️📂Core kernel Yaiwes/Hivemind/sdk-proxy/error-classifier.js
// Single classification point for every provider failure the proxy can see.
//
// The distinction that matters is not "what went wrong" but "will dialling
// again ever help". A 400 "You're out of extra usage" is a human-actionable
// permanent failure: retrying it opens a fresh TCP connection per attempt and
// burns the host's ephemeral port pool for nothing. Callers must be able to
// tell that apart from a 503 without string-matching, so every error leaves
// this module as { retryable, reason, status }.

const QUOTA_PATTERNS = [
  /out of extra usage/i,
  /credit balance is too low/i,
  /insufficient[_\s-]?quota/i,
  /billing[_\s-]?(?:hard[_\s-]?)?limit/i,
  /purchase (?:more )?credits/i,
  /add more at claude\.ai\/settings\/usage/i,
];

const AUTH_PATTERNS = [
  /invalid[_\s-]?api[_\s-]?key/i,
  /authentication[_\s-]?error/i,
  /oauth token (?:has )?expired/i,
  /invalid bearer token/i,
  /unauthorized/i,
];

const MODEL_NOT_FOUND_PATTERNS = [
  /model[:\s].*not[_\s-]?found/i,
  /not_found_error/i,
  /unknown model/i,
];

// Local resource exhaustion. Deliberately NOT retryable: EADDRNOTAVAIL means
// the kernel has no ephemeral port left, and every retry makes that worse.
// This is the exact failure mode of the 2026-08-24 host outage.
const LOCAL_EXHAUSTION_PATTERNS = [
  /EADDRNOTAVAIL/,
  /can't assign requested address/i,
  /EMFILE/,
  /ENFILE/,
  /too many open files/i,
];

const NETWORK_PATTERNS = [
  /ECONNRESET/, /ECONNREFUSED/, /EPIPE/, /ENOTFOUND/, /EAI_AGAIN/,
  /ETIMEDOUT/, /ESOCKETTIMEDOUT/, /socket hang up/i,
  /network[_\s-]?error/i, /connection error/i,
];

const OVERLOAD_PATTERNS = [
  /overloaded/i, /rate[_\s-]?limit/i, /too many requests/i,
];

/** Reasons that must never be retried against the same credential. */
export const PERMANENT_REASONS = Object.freeze([
  "quota_exhausted",
  "auth_invalid",
  "forbidden",
  "model_not_found",
  "invalid_request",
  "request_too_large",
  "local_port_exhaustion",
  "unknown",
]);

/** Reasons that count toward opening the per-credential circuit. */
export const CIRCUIT_REASONS = Object.freeze([
  "quota_exhausted",
  "auth_invalid",
  "forbidden",
  "local_port_exhaustion",
]);

const matchesAny = (patterns, text) => patterns.some((p) => p.test(text));

/**
 * Recover an HTTP status that got flattened into a text blob.
 *
 * The OAuth path fails as `Claude Code process exited with code 1`, which
 * erases the underlying status. The real cause is only visible in the
 * subprocess's stderr, in shapes like `API Error: 400 {"type":"error",...}`
 * or `Anthropic API error (429)`.
 */
export function parseStatusFromText(text) {
  if (!text) return null;
  const patterns = [
    /API Error:?\s*(\d{3})\b/i,
    /\berror\s*\((\d{3})\)/i,
    /\bstatus(?:\s*code)?[:\s]+(\d{3})\b/i,
    /\bHTTP\s+(\d{3})\b/i,
    /"status"\s*:\s*(\d{3})\b/,
  ];
  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) {
      const status = Number(match[1]);
      if (status >= 400 && status <= 599) return status;
    }
  }
  return null;
}

function reasonForStatus(status, text) {
  switch (true) {
    case status === 400:
      return matchesAny(QUOTA_PATTERNS, text) ? "quota_exhausted" : "invalid_request";
    case status === 401:
      return "auth_invalid";
    case status === 402:
      return "quota_exhausted";
    case status === 403:
      return "forbidden";
    case status === 404:
      return "model_not_found";
    case status === 408:
      return "timeout";
    case status === 409:
      return "conflict";
    case status === 413:
      return "request_too_large";
    case status === 422:
      return "invalid_request";
    case status === 429:
      return "rate_limited";
    case status >= 500 && status <= 599:
      return "server_error";
    default:
      return null;
  }
}

function reasonFromText(text) {
  // Order matters: local exhaustion and quota are checked before the generic
  // network patterns, because their messages often contain both.
  if (matchesAny(LOCAL_EXHAUSTION_PATTERNS, text)) return "local_port_exhaustion";
  if (matchesAny(QUOTA_PATTERNS, text)) return "quota_exhausted";
  if (matchesAny(AUTH_PATTERNS, text)) return "auth_invalid";
  if (matchesAny(MODEL_NOT_FOUND_PATTERNS, text)) return "model_not_found";
  if (matchesAny(OVERLOAD_PATTERNS, text)) return "rate_limited";
  if (matchesAny(NETWORK_PATTERNS, text)) return "network_error";
  return null;
}

function statusForReason(reason, fallback) {
  if (fallback) return fallback;
  const map = {
    quota_exhausted: 402,
    auth_invalid: 401,
    forbidden: 403,
    model_not_found: 404,
    invalid_request: 400,
    request_too_large: 413,
    rate_limited: 429,
    timeout: 504,
    conflict: 409,
    server_error: 502,
    network_error: 502,
    local_port_exhaustion: 503,
    circuit_open: 503,
    load_shed: 429,
  };
  return map[reason] || 500;
}

function parseRetryAfterMs(err) {
  const header = err?.headers?.["retry-after"] ?? err?.responseHeaders?.["retry-after"];
  if (header == null) return null;
  const seconds = Number(header);
  return Number.isFinite(seconds) && seconds >= 0 ? Math.round(seconds * 1000) : null;
}

/**
 * Classify a provider failure.
 *
 * @param {unknown} err          the thrown error
 * @param {object}  [context]
 * @param {string}  [context.stderr]  captured subprocess stderr (OAuth path)
 * @returns {{retryable: boolean, reason: string, status: number,
 *            message: string, retryAfterMs: number|null}}
 */
export function classifyError(err, context = {}) {
  const message = String(err?.message ?? err ?? "Unknown error");
  const stderr = String(context.stderr || "");
  const haystack = `${message}\n${stderr}`;

  // Errors we raised ourselves already carry an authoritative verdict
  // (circuit open, load shed). Never re-derive it from their HTTP status —
  // a 503 from an open circuit is emphatically not a retryable server error.
  if (typeof err?.reason === "string" && typeof err?.retryable === "boolean") {
    return {
      retryable: err.retryable,
      reason: err.reason,
      status: Number.isInteger(err.status) ? err.status : statusForReason(err.reason, null),
      message,
      retryAfterMs: Number.isInteger(err.retryAfterMs) ? err.retryAfterMs : parseRetryAfterMs(err),
    };
  }

  // An explicit numeric status on the error object is the strongest signal;
  // otherwise dig it back out of the text the subprocess left behind.
  const explicitStatus =
    Number.isInteger(err?.status) ? err.status :
    Number.isInteger(err?.statusCode) ? err.statusCode :
    parseStatusFromText(haystack);

  const reason =
    (explicitStatus && reasonForStatus(explicitStatus, haystack)) ||
    reasonFromText(haystack) ||
    "unknown";

  return {
    retryable: !PERMANENT_REASONS.includes(reason),
    reason,
    status: statusForReason(reason, explicitStatus),
    message,
    retryAfterMs: parseRetryAfterMs(err),
  };
}

/** Should this classification count toward opening the credential's circuit? */
export function opensCircuit(classification) {
  return CIRCUIT_REASONS.includes(classification.reason);
}

/** Wire format shared with the Rails side (Providers::ErrorClassifier). */
export function toErrorBody(classification) {
  return {
    error: { message: classification.message, type: classification.reason },
    retryable: classification.retryable,
    reason: classification.reason,
    ...(classification.retryAfterMs != null
      ? { retry_after_ms: classification.retryAfterMs }
      : {}),
  };
}
